tests/ImageSharp.Tests/Processing/ImageProcessingContextTests.cs

```diff
@@ -0,0 +1,170 @@
+
+        private readonly Mock<ICloningImageProcessor<Rgba32>> cloningProcessorImpl;
+
+        private static readonly Rectangle Bounds = new Rectangle(3, 3, 5, 5);
+
+        public ImageProcessingContextTests()
+        {
+            this.processorDefinition = new Mock<IImageProcessor>();
+            this.regularProcessorImpl = new Mock<IImageProcessor<Rgba32>>();
+            this.cloningProcessorImpl = new Mock<ICloningImageProcessor<Rgba32>>();
+        }
+
+        // bool throwException, bool useBounds
+        public static readonly TheoryData<bool, bool> ProcessorTestData = new TheoryData<bool, bool>()
+                                                                          {
+                                                                              { false, false },
+                                                                              { false, true },
+                                                                              { true, false },
+                                                                              { true, true }
+                                                                          };
+        [Theory]
+        [MemberData(nameof(ProcessorTestData))]
+        public void Mutate_RegularProcessor(bool throwException, bool useBounds)
+        {
+            this.SetupRegularProcessor(throwException);
+
+            if (throwException)
+            {
+                Assert.Throws<ImageProcessingException>(() => this.MutateApply(useBounds));
+            }
+            else
+            {
+                this.MutateApply(useBounds);
+            }
+
+            this.regularProcessorImpl.Verify(p => p.Apply(), Times.Once());
+            this.regularProcessorImpl.Verify(p => p.Dispose(), Times.Once());
+        }
+
+        [Theory]
+        [MemberData(nameof(ProcessorTestData))]
+        public void Clone_RegularProcessor(bool throwException, bool useBounds)
+        {
+            this.SetupRegularProcessor(throwException);
+
+            if (throwException)
+            {
+                Assert.Throws<ImageProcessingException>(() => this.CloneApply(useBounds));
+            }
+            else
+            {
+                this.CloneApply(useBounds);
+            }
+
+            this.regularProcessorImpl.Verify(p => p.Apply(), Times.Once());
+            this.regularProcessorImpl.Verify(p => p.Dispose(), Times.Once());
+        }
+
+        [Theory]
+        [MemberData(nameof(ProcessorTestData))]
+        public void Mutate_CloningProcessor(bool throwException, bool useBounds)
+        {
+            this.SetupCloningProcessor(throwException);
+
+            if (throwException)
+            {
+                Assert.Throws<ImageProcessingException>(() => this.MutateApply(useBounds));
+            }
+            else
+            {
+                this.MutateApply(useBounds);
+            }
+
+            this.cloningProcessorImpl.Verify(p => p.CloneAndApply(), Times.Once());
```

Comment:
@JimBobSquarePants oops. We should expect `this.cloningProcessorImpl.Verify(p => p.Apply(), Times.Once());` here instead. To avoid conflicts, I'll leave it to you.